org.adempiere.base/src/org/compiere/model/MProcessDrillRulePara.java

```diff
@@ -95,4 +97,36 @@ public PO markImmutable() {
 	public void setParentID(int id) {
 		setAD_Process_DrillRule_ID(id);
 	}
+	
+	@Override
+	protected boolean afterSave(boolean newRecord, boolean success) {
+		if(success) {
+			String whereClause = MProcessPara.COLUMNNAME_AD_Process_Para_ID + " = ? ";
+			MProcessPara processPara = new Query(getCtx(), MProcessPara.Table_Name, whereClause, get_TrxName()).setParameters(getAD_Process_Para_ID()).first();
+			boolean isValid = (processPara.isRange() && (!Util.isEmpty(getParameterDefault()) || !Util.isEmpty(getParameterToDefault()))) ||
+					((!processPara.isRange() && !Util.isEmpty(getParameterDefault())));
+			if(processPara.isMandatory()) {
+				validateParent(new Object[] {isValid ? "Y" : "N", getAD_Process_DrillRule_ID()});
+			}
+		}
+		return super.afterSave(newRecord, success);
+	}
+	
+	@Override
+	protected boolean afterDelete (boolean success) {
+		if(success) {
+			String whereClause = MProcessPara.COLUMNNAME_AD_Process_Para_ID + " = ? ";
+			MProcessPara processPara = new Query(getCtx(), MProcessPara.Table_Name, whereClause, get_TrxName()).setParameters(getAD_Process_Para_ID()).first();
+			if(processPara.isMandatory()) {
+				validateParent(new Object[] {"N", getAD_Process_DrillRule_ID()});
+			}
+		}
+		return super.afterDelete(success);
+	}
+	
+	private void validateParent(Object[] para) {
+			String sql = " UPDATE " + MProcessDrillRule.Table_Name + " SET " + MProcessDrillRule.COLUMNNAME_IsValid + " = ? "
```

Comment:
here too - this is better done using the model class instead of direct SQL

Reply:
hi @CarlosRuiz-globalqss , using the model class in afterDelete causes the following issue when deleting Drill Rule Para:

![Screenshot 2023-01-16 at 13 02 28](https://user-images.githubusercontent.com/93127072/212674199-ed730372-7176-4357-85e7-c60a0c226bff.png)

Since we don't want to validate and update the parent if the deletion was not successful, the validation probably needs to be in AfterDelete. Is there a solution to avoid getting this warning message?

Reply:
From ux perspective this is a phantom error message. If no other solution then probably  would be better validate it only by process not after save, after delete. wdyt ?

Reply:
Thanks @norbertbede - I understand your concern about ux perspective.

From technical perspective, it seems bad to use direct SQL instead of model to overcome a ux problem, in such case is better to solve the ux problem.

You would need to check where is the message triggered and check for some condition to avoid it when is controlled (like in the case of payment term and here)

Reply:
Hi @CarlosRuiz-globalqss ,

With @norbertbede we decided to remove the validation from Drill Rule and Drill Rule Para beforeSave/afterSave/afterDelete, so the only way to validate the Drill Rule will be through the new process for now.

The issue regarding to the warning message could go probably for another ticket.

If you agree with this, we would update the ticket description too and it could be closed sooner.

Reply:
> With @norbertbede we decided to remove the validation from Drill Rule and Drill Rule Para beforeSave/afterSave/afterDelete, so the only way to validate the Drill Rule will be through the new process for now.
> 
> The issue regarding to the warning message could go probably for another ticket.
> 
> If you agree with this, we would update the ticket description too and it could be closed sooner.

I did the following test:

- Create _Drill Rule_ with _Process_ = Accounting Fact Daily_RV_Fact_Acct_Day
- _Capture Process Parameters_ - define _Accounting Schema_ which is mandatory
- Run _Validate_ process
- The rule is set as Valid because the mandatory parameter is set
- Delete the _Accounting Schema_ record
- The rule is still valid, but it will fail if used - which is what you're trying to avoid with this ticket

So, which one is worst for user: a harmless phantom error message?  or allowing to use an invalid rule?

Reply:
Hi @CarlosRuiz-globalqss, I've tested similar use case in Info Column - when I delete an Info Column it also calls the validation method of the parent, which sets the IsValid column, then saves the parent record. In this case however there is no phantom error.

I could not find how is it solved there, or what is the difference between the two implementations. While debugging, I've found that in GridTable:3991, method hasChanged(PO po) it checks the current value of "Updated" with the old value of "Updated". When I'm on the info column, they are the same, but when I'm on drill rule para, they are different. Seems like the Updated column of the info window is not changed when I set IsValid after deleting an info column? I could not get further than this.